cmd/snap-repair/trace.go

```diff
@@ -0,0 +1,175 @@
+func (rt *repairTrace) Repair() string {
+	seq := filepath.Base(filepath.Dir(rt.path))
+	brand := filepath.Base(filepath.Dir(filepath.Dir(rt.path)))
+
+	return fmt.Sprintf("%s-%s", brand, seq)
+}
+
+// Revision returns the revision of the repair
+func (rt *repairTrace) Revision() string {
+	return revFromFilepath(rt.path)
+}
+
+// Summary returns the summary of the repair that was run
+func (rt *repairTrace) Summary() string {
+	f, err := os.Open(rt.path)
+	if err != nil {
+		return "cannot read summary"
+	}
+	defer f.Close()
+
+	needle := "summary: "
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		s := scanner.Text()
+		if strings.HasPrefix(s, needle) {
+			return s[len(needle):]
+		}
+	}
+
+	return "cannot find summary"
+}
+
+// Status returns the status of the given repair {done,skip,retry,running}
+func (rt *repairTrace) Status() string {
+	return filepath.Ext(rt.path)[1:]
+}
+
+func indentPrefix(level int) string {
+	return strings.Repeat(" ", level)
+}
+
+// WriteScriptIndented outputs the script that produced this repair output
+// to the given writer w with the indent level given by indent.
+func (rt *repairTrace) WriteScriptIndented(w io.Writer, indent int) {
+	scriptPath := rt.path[:strings.LastIndex(rt.path, ".")] + ".script"
+	f, err := os.Open(scriptPath)
+	if err != nil {
+		fmt.Fprintf(w, "cannot read script: %v", err)
+		return
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		fmt.Fprintf(w, "%s%s\n", indentPrefix(indent), scanner.Text())
+	}
+	if scanner.Err() != nil {
+		fmt.Fprintf(w, "%serror: %s\n", indentPrefix(indent), scanner.Err())
+	}
+}
+
+// WriteOutputIndented outputs the repair output to the given writer w
+// with the indent level given by indent.
+func (rt *repairTrace) WriteOutputIndented(w io.Writer, indent int) {
+	f, err := os.Open(rt.path)
+	if err != nil {
+		fmt.Fprintf(w, "  error: %s\n", err)
+		return
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	// move forward in the log to where the actual script output starts
+	for scanner.Scan() {
+		if scanner.Text() == "output:" {
+			break
+		}
+	}
+	// write the script output to w
+	for scanner.Scan() {
+		fmt.Fprintf(w, "%s%s\n", indentPrefix(indent), scanner.Text())
+	}
+	if scanner.Err() != nil {
+		fmt.Fprintf(w, "%serror: %s\n", indentPrefix(indent), scanner.Err())
+	}
+}
+
+// validSuffix returns true if the given traceName is something repairTrace
+// understands.
+func (rt *repairTrace) validSuffix(traceName string) bool {
+	for _, valid := range []string{".retry", ".skip", ".done", ".running"} {
+		if strings.HasSuffix(traceName, valid) {
+			return true
+		}
+	}
+	return false
+}
+
+// revFromFilepath is a helper that extrace the revision number from the
+// filename of the repairTrace
+func revFromFilepath(name string) string {
+	var rev int
+	if _, err := fmt.Sscanf(filepath.Base(name), "r%d.", &rev); err == nil {
+		return strconv.Itoa(rev)
+	}
+	return "?"
```

Comment:
This sounds surprising and I'm afraid of ending up with "?" in curious and potentially dangerous places. Similar to above, I think it should be responsibility of the call site to decide what to do when it goes wrong.